apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -0,0 +1,43 @@
+import asyncHandler = require('express-async-handler');
+import * as sqldb from '@prairielearn/postgres';
+import { IdSchema } from '../lib/db-types';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+export default asyncHandler(async (req, res, next) => {
+  const userSessionId = await sqldb.queryOptionalRow(
+    sql.select_session_id,
+    { session_id: req.session.id, user_id: res.locals.user.user_id },
+    IdSchema,
+  );
+  const params = {
+    ip_address: req.ip,
+    user_id: res.locals.authn_user.user_id,
+    user_session_id: userSessionId,
+    user_agent: req.headers['user-agent'],
+    accept_language: req.headers['accept-language'],
+  };
+
+  let client_fingerprint_id = await sqldb.queryOptionalRow(
+    sql.select_client_fingerprint,
+    params,
+    IdSchema,
+  );
+
+  if (!client_fingerprint_id) {
+    client_fingerprint_id = await sqldb.queryRow(sql.insert_client_fingerprint, params, IdSchema);
+  }
+  if (
+    res.locals.assessment_instance &&
+    res.locals.assessment_instance?.last_client_fingerprint_id.toString() !== client_fingerprint_id
```

Comment:
Shouldn't need `toString()` here -- these should already all explicitly be typed as strings?

Reply:
These are typed as numbers. I havenow  updated this to use `IdSchema.parse()` instead of `toString()` here which will explicitly take a non negative integer and parse as a string. Will that work instead?

Reply:
Oh good point, I forgot that this data is coming out of `res.locals` which doesn't have any sensible types in it. Using IdSchema here is a good idea.